Flutter 3.29 Fatal crash with java.lang.AssertionError

### Steps to reproduce

1 Create new android project
`flutter create --org com.example --platforms=android -a kotlin hello`

2 Run on Android device with a physical keyboard, example `Urovo RT40`

3 Press and hold the `ALT` button
`PhysicalKeyboardKey#91568(usbHidUsage: "0x1100000039", debugName: "Key with ID 0x1100000039")`


### Expected results

nothing

### Actual results

Crash

### Code sample

<details><summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});
  
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;

  void _incrementCounter() {
    setState(() {
      _counter++;
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        backgroundColor: Theme.of(context).colorScheme.inversePrimary,
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            const Text('You have pushed the button this many times:'),
            Text(
              '$_counter',
              style: Theme.of(context).textTheme.headlineMedium,
            ),
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _incrementCounter,
        tooltip: 'Increment',
        child: const Icon(Icons.add),
      ),
    );
  }
}

```

</details>


### Screenshots or Video

<details open>
<summary>UROVO RT40</summary>
![Image](https://github.com/user-attachments/assets/b45c83a7-5361-42fd-986f-2c929c1c5c27)

https://github.com/user-attachments/assets/a6d5108e-378f-48bb-8fb9-93d9a94f1365

</details>


### Logs

<details ><summary>Logs</summary>

```console
E/InputEventSender( 4705): Exception dispatching finished signal.
E/MessageQueue-JNI( 4705): Exception in MessageQueue callback: handleReceiveCallback
E/MessageQueue-JNI( 4705): java.lang.AssertionError: The key was empty
E/MessageQueue-JNI( 4705): 	at io.flutter.embedding.android.KeyEmbedderResponder.updatePressingState(KeyEmbedderResponder.java:116)
E/MessageQueue-JNI( 4705): 	at io.flutter.embedding.android.KeyEmbedderResponder.synthesizeEvent(KeyEmbedderResponder.java:390)
E/MessageQueue-JNI( 4705): 	at io.flutter.embedding.android.KeyEmbedderResponder.lambda$synchronizePressingKey$1$io-flutter-embedding-android-KeyEmbedderResponder(KeyEmbedderResponder.java:185)
E/MessageQueue-JNI( 4705): 	at io.flutter.embedding.android.KeyEmbedderResponder$$ExternalSyntheticLambda2.run(D8$$SyntheticClass:0)
E/MessageQueue-JNI( 4705): 	at io.flutter.embedding.android.KeyEmbedderResponder.handleEventImpl(KeyEmbedderResponder.java:375)
E/MessageQueue-JNI( 4705): 	at io.flutter.embedding.android.KeyEmbedderResponder.handleEvent(KeyEmbedderResponder.java:431)
E/MessageQueue-JNI( 4705): 	at io.flutter.embedding.android.KeyboardManager.handleEvent(KeyboardManager.java:229)
E/MessageQueue-JNI( 4705): 	at io.flutter.embedding.android.FlutterView.dispatchKeyEvent(FlutterView.java:881)
3
E/MessageQueue-JNI( 4705): 	at android.view.ViewGroup.dispatchKeyEvent(ViewGroup.java:1922)
E/MessageQueue-JNI( 4705): 	at com.android.internal.policy.DecorView.superDispatchKeyEvent(DecorView.java:453)
E/MessageQueue-JNI( 4705): 	at com.android.internal.policy.PhoneWindow.superDispatchKeyEvent(PhoneWindow.java:1839)
E/MessageQueue-JNI( 4705): 	at android.app.Activity.dispatchKeyEvent(Activity.java:3972)
E/MessageQueue-JNI( 4705): 	at com.android.internal.policy.DecorView.dispatchKeyEvent(DecorView.java:367)
```

</details>


### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
[✓] Flutter (Channel stable, 3.29.0, on Ubuntu 24.04.2 LTS 6.8.0-54-generic, locale ru_RU.UTF-8) [86ms]
    • Flutter version 3.29.0 on channel stable at /usr/local/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 35c388afb5 (3 недели назад), 2025-02-10 12:48:41 -0800
    • Engine revision f73bfc4522
    • Dart version 3.7.0
    • DevTools version 2.42.2

[✓] Android toolchain - develop for Android devices (Android SDK version 35.0.0) [1 094ms]
    • Android SDK at /home/pavel/Android/Sdk
    • Platform android-35, build-tools 35.0.0
    • Java binary at: /usr/local/android-studio/jbr/bin/java
      This is the JDK bundled with the latest Android Studio installation on this machine.
      To manually set the JDK path, use: `flutter config --jdk-dir="path/to/jdk"`.
    • Java version OpenJDK Runtime Environment (build 21.0.4+-12422083-b607.1)
    • All Android licenses accepted.

[✓] Chrome - develop for the web [13ms]
    • Chrome at google-chrome

[✓] Linux toolchain - develop for Linux desktop [140ms]
    • Ubuntu clang version 18.1.3 (1ubuntu1)
    • cmake version 3.28.3
    • ninja version 1.11.1
    • pkg-config version 1.8.1

[✓] Android Studio (version 2024.2) [11ms]
    • Android Studio at /usr/local/android-studio
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 21.0.4+-12422083-b607.1)

[✓] VS Code (version 1.97.2) [10ms]
    • VS Code at /usr/share/code
    • Flutter extension version 3.106.0

[✓] Connected device (3 available) [190ms]
    • RT40 (mobile)   • 4932d4d1 • android-arm64  • Android 10 (API 29)
    • Linux (desktop) • linux    • linux-x64      • Ubuntu 24.04.2 LTS 6.8.0-54-generic
    • Chrome (web)    • chrome   • web-javascript • Google Chrome 133.0.6943.141

[✓] Network resources [1 390ms]
    • All expected network resources are available.

• No issues found!
```

</details>

> To enable the logs set the `debugPrintKeyboardEvents` to `true`:
> 
> void main() {
>   debugPrintKeyboardEvents = true;
>   runApp(const MyApp());
> }

<details>
<summary>Logs</summary>

```console
<!-- short press on the button -->
I/flutter ( 5200): KEYBOARD: Key event received: KeyDownEvent#2c73e(physicalKey: PhysicalKeyboardKey#91568(usbHidUsage: "0x1100000039", debugName: "Key with ID 0x1100000039"), logicalKey: LogicalKeyboardKey#e5ce6(keyId: "0x200000104", keyLabel: "Alt Left", debugName: "Alt Left"), character: null, timeStamp: 0:00:00.560994)
I/flutter ( 5200): KEYBOARD: Pressed state before processing the event:
I/flutter ( 5200):     Empty
I/flutter ( 5200): KEYBOARD: Pressed state after processing the event:
I/flutter ( 5200):     PhysicalKeyboardKey#91568(usbHidUsage: "0x1100000039", debugName: "Key with ID 0x1100000039"): LogicalKeyboardKey#e5ce6(keyId: "0x200000104", keyLabel: "Alt Left", debugName: "Alt Left")
I/flutter ( 5200): KEYBOARD: Key event received: KeyUpEvent#ab477(physicalKey: PhysicalKeyboardKey#91568(usbHidUsage: "0x1100000039", debugName: "Key with ID 0x1100000039"), logicalKey: LogicalKeyboardKey#e5ce6(keyId: "0x200000104", keyLabel: "Alt Left", debugName: "Alt Left"), character: null, timeStamp: 0:00:00.560994, synthesized)
I/flutter ( 5200): KEYBOARD: Pressed state before processing the event:
I/flutter ( 5200):     PhysicalKeyboardKey#91568(usbHidUsage: "0x1100000039", debugName: "Key with ID 0x1100000039"): LogicalKeyboardKey#e5ce6(keyId: "0x200000104", keyLabel: "Alt Left", debugName: "Alt Left")
I/flutter ( 5200): KEYBOARD: Pressed state after processing the event:
I/flutter ( 5200):     Empty

<!-- long press on the button -->

I/flutter ( 5200): KEYBOARD: Key event received: KeyDownEvent#b680c(physicalKey: PhysicalKeyboardKey#91568(usbHidUsage: "0x1100000039", debugName: "Key with ID 0x1100000039"), logicalKey: LogicalKeyboardKey#e5ce6(keyId: "0x200000104", keyLabel: "Alt Left", debugName: "Alt Left"), character: null, timeStamp: 0:00:00.732393)
I/flutter ( 5200): KEYBOARD: Pressed state before processing the event:
I/flutter ( 5200):     Empty
I/flutter ( 5200): KEYBOARD: Pressed state after processing the event:
I/flutter ( 5200):     PhysicalKeyboardKey#91568(usbHidUsage: "0x1100000039", debugName: "Key with ID 0x1100000039"): LogicalKeyboardKey#e5ce6(keyId: "0x200000104", keyLabel: "Alt Left", debugName: "Alt Left")
I/flutter ( 5200): KEYBOARD: Key event received: KeyUpEvent#a392a(physicalKey: PhysicalKeyboardKey#91568(usbHidUsage: "0x1100000039", debugName: "Key with ID 0x1100000039"), logicalKey: LogicalKeyboardKey#e5ce6(keyId: "0x200000104", keyLabel: "Alt Left", debugName: "Alt Left"), character: null, timeStamp: 0:00:00.732393, synthesized)
I/flutter ( 5200): KEYBOARD: Pressed state before processing the event:
I/flutter ( 5200):     PhysicalKeyboardKey#91568(usbHidUsage: "0x1100000039", debugName: "Key with ID 0x1100000039"): LogicalKeyboardKey#e5ce6(keyId: "0x200000104", keyLabel: "Alt Left", debugName: "Alt Left")
I/flutter ( 5200): KEYBOARD: Pressed state after processing the event:
I/flutter ( 5200):     Empty
E/InputEventSender( 5200): Exception dispatching finished signal.
E/MessageQueue-JNI( 5200): Exception in MessageQueue callback: handleReceiveCallback
E/MessageQueue-JNI( 5200): java.lang.AssertionError: The key was empty
E/MessageQueue-JNI( 5200): 	at io.flutter.embedding.android.KeyEmbedderResponder.updatePressingState(KeyEmbedderResponder.java:116)
E/MessageQueue-JNI( 5200): 	at io.flutter.embedding.android.KeyEmbedderResponder.synthesizeEvent(KeyEmbedderResponder.java:390)
E/MessageQueue-JNI( 5200): 	at io.flutter.embedding.android.KeyEmbedderResponder.lambda$synchronizePressingKey$1$io-flutter-embedding-android-KeyEmbedderResponder(KeyEmbedderResponder.java:185)
E/MessageQueue-JNI( 5200): 	at io.flutter.embedding.android.KeyEmbedderResponder$$ExternalSyntheticLambda2.run(D8$$SyntheticClass:0)
E/MessageQueue-JNI( 5200): 	at io.flutter.embedding.android.KeyEmbedderResponder.handleEventImpl(KeyEmbedderResponder.java:375)
E/MessageQueue-JNI( 5200): 	at io.flutter.embedding.android.KeyEmbedderResponder.handleEvent(KeyEmbedderResponder.java:431)
E/MessageQueue-JNI( 5200): 	at io.flutter.embedding.android.KeyboardManager.handleEvent(KeyboardManager.java:229)
E/MessageQueue-JNI( 5200): 	at io.flutter.embedding.android.FlutterView.dispatchKeyEvent(FlutterView.java:881)
E/MessageQueue-JNI( 5200): 	at android.view.ViewGroup.dispatchKeyEvent(ViewGroup.java:1922)
E/MessageQueue-JNI( 5200): 	at android.view.ViewGroup.dispatchKeyEvent(ViewGroup.java:1922)
E/MessageQueue-JNI( 5200): 	at android.view.ViewGroup.dispatchKeyEvent(ViewGroup.java:1922)
E/MessageQueue-JNI( 5200): 	at com.android.internal.policy.DecorView.superDispatchKeyEvent(DecorView.java:453)
E/MessageQueue-JNI( 5200): 	at com.android.internal.policy.PhoneWindow.superDispatchKeyEvent(PhoneWindow.java:1839)
E/MessageQueue-JNI( 5200): 	at android.app.Activity.dispatchKeyEvent(Activity.java:3972)
E/MessageQueue-JNI( 5200): 	at com.android.internal.policy.DecorView.dispatchKeyEvent(DecorView.java:367)
E/MessageQueue-JNI( 5200): 	at android.view.ViewRootImpl$ViewPostImeInputStage.processKeyEvent(ViewRootImpl.java:5944)
E/MessageQueue-JNI( 5200): 	at android.view.ViewRootImpl$ViewPostImeInputStage.onProcess(ViewRootImpl.java:5812)
E/MessageQueue-JNI( 5200): 	at android.view.ViewRootImpl$InputStage.deliver(ViewRootImpl.java:5319)
E/MessageQueue-JNI( 5200): 	at android.view.ViewRootImpl$InputStage.onDeliverToNext(ViewRootImpl.java:5372)
E/MessageQueue-JNI( 5200): 	at android.view.ViewRootImpl$InputStage.forward(ViewRootImpl.java:5338)
E/MessageQueue-JNI( 5200): 	at android.view.ViewRootImpl$AsyncInputStage.forward(ViewRootImpl.java:5478)
E/MessageQueue-JNI( 5200): 	at android.view.ViewRootImpl$InputStage.apply(ViewRootImpl.java:5346)
E/MessageQueue-JNI( 5200): 	at android.view.ViewRootImpl$AsyncInputStage.apply(ViewRootImpl.java:5535)
E/MessageQueue-JNI( 5200): 	at android.view.ViewRootImpl$InputStage.deliver(ViewRootImpl.java:5319)
E/MessageQueue-JNI( 5200): 	at android.view.ViewRootImpl$InputStage.onDeliverToNext(ViewRootImpl.java:5372)
E/MessageQueue-JNI( 5200): 	at android.view.ViewRootImpl$InputStage.forward(ViewRootImpl.java:5338)
E/MessageQueue-JNI( 5200): 	at android.view.ViewRootImpl$InputStage.apply(ViewRootImpl.java:5346)
E/MessageQueue-JNI( 5200): 	at android.view.ViewRootImpl$InputStage.deliver(ViewRootImpl.java:5319)
E/MessageQueue-JNI( 5200): 	at android.view.ViewRootImpl$InputStage.onDeliverToNext(ViewRootImpl.java:5372)
E/MessageQueue-JNI( 5200): 	at android.view.ViewRootImpl$InputStage.forward(ViewRootImpl.java:5338)
E/MessageQueue-JNI( 5200): 	at android.view.ViewRootImpl$AsyncInputStage.forward(ViewRootImpl.java:5511)
E/MessageQueue-JNI( 5200): 	at android.view.ViewRootImpl$ImeInputStage.onFinishedInputEvent(ViewRootImpl.java:5672)
E/MessageQueue-JNI( 5200): 	at android.view.inputmethod.InputMethodManager$PendingEvent.run(InputMethodManager.java:3065)
E/MessageQueue-JNI( 5200): 	at android.view.inputmethod.InputMethodManager.invokeFinishedInputEventCallback(InputMethodManager.java:2608)
E/MessageQueue-JNI( 5200): 	at android.view.inputmethod.InputMethodManager.finishedInputEvent(InputMethodManager.java:2599)
E/MessageQueue-JNI( 5200): 	at android.view.inputmethod.InputMethodManager$ImeInputEventSender.onInputEventFinished(InputMethodManager.java:3042)
E/MessageQueue-JNI( 5200): 	at android.view.InputEventSender.dispatchInputEventFinished(InputEventSender.java:143)
E/MessageQueue-JNI( 5200): 	at android.os.MessageQueue.nativePollOnce(Native Method)
E/MessageQueue-JNI( 5200): 	at android.os.MessageQueue.next(MessageQueue.java:336)
E/MessageQueue-JNI( 5200): 	at android.os.Looper.loop(Looper.java:174)
E/MessageQueue-JNI( 5200): 	at android.app.ActivityThread.main(ActivityThread.java:7397)
E/MessageQueue-JNI( 5200): 	at java.lang.reflect.Method.invoke(Native Method)
E/MessageQueue-JNI( 5200): 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
E/MessageQueue-JNI( 5200): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:935)
D/AndroidRuntime( 5200): Shutting down VM
E/AndroidRuntime( 5200): FATAL EXCEPTION: main
E/AndroidRuntime( 5200): Process: com.example.hello, PID: 5200
E/AndroidRuntime( 5200): java.lang.AssertionError: The key was empty
E/AndroidRuntime( 5200): 	at io.flutter.embedding.android.KeyEmbedderResponder.updatePressingState(KeyEmbedderResponder.java:116)
E/AndroidRuntime( 5200): 	at io.flutter.embedding.android.KeyEmbedderResponder.synthesizeEvent(KeyEmbedderResponder.java:390)
E/AndroidRuntime( 5200): 	at io.flutter.embedding.android.KeyEmbedderResponder.lambda$synchronizePressingKey$1$io-flutter-embedding-android-KeyEmbedderResponder(KeyEmbedderResponder.java:185)
E/AndroidRuntime( 5200): 	at io.flutter.embedding.android.KeyEmbedderResponder$$ExternalSyntheticLambda2.run(D8$$SyntheticClass:0)
E/AndroidRuntime( 5200): 	at io.flutter.embedding.android.KeyEmbedderResponder.handleEventImpl(KeyEmbedderResponder.java:375)
E/AndroidRuntime( 5200): 	at io.flutter.embedding.android.KeyEmbedderResponder.handleEvent(KeyEmbedderResponder.java:431)
E/AndroidRuntime( 5200): 	at io.flutter.embedding.android.KeyboardManager.handleEvent(KeyboardManager.java:229)
E/AndroidRuntime( 5200): 	at io.flutter.embedding.android.FlutterView.dispatchKeyEvent(FlutterView.java:881)
E/AndroidRuntime( 5200): 	at android.view.ViewGroup.dispatchKeyEvent(ViewGroup.java:1922)
E/AndroidRuntime( 5200): 	at android.view.ViewGroup.dispatchKeyEvent(ViewGroup.java:1922)
E/AndroidRuntime( 5200): 	at android.view.ViewGroup.dispatchKeyEvent(ViewGroup.java:1922)
E/AndroidRuntime( 5200): 	at com.android.internal.policy.DecorView.superDispatchKeyEvent(DecorView.java:453)
E/AndroidRuntime( 5200): 	at com.android.internal.policy.PhoneWindow.superDispatchKeyEvent(PhoneWindow.java:1839)
E/AndroidRuntime( 5200): 	at android.app.Activity.dispatchKeyEvent(Activity.java:3972)
E/AndroidRuntime( 5200): 	at com.android.internal.policy.DecorView.dispatchKeyEvent(DecorView.java:367)
E/AndroidRuntime( 5200): 	at android.view.ViewRootImpl$ViewPostImeInputStage.processKeyEvent(ViewRootImpl.java:5944)
E/AndroidRuntime( 5200): 	at android.view.ViewRootImpl$ViewPostImeInputStage.onProcess(ViewRootImpl.java:5812)
E/AndroidRuntime( 5200): 	at android.view.ViewRootImpl$InputStage.deliver(ViewRootImpl.java:5319)
E/AndroidRuntime( 5200): 	at android.view.ViewRootImpl$InputStage.onDeliverToNext(ViewRootImpl.java:5372)
E/AndroidRuntime( 5200): 	at android.view.ViewRootImpl$InputStage.forward(ViewRootImpl.java:5338)
E/AndroidRuntime( 5200): 	at android.view.ViewRootImpl$AsyncInputStage.forward(ViewRootImpl.java:5478)
E/AndroidRuntime( 5200): 	at android.view.ViewRootImpl$InputStage.apply(ViewRootImpl.java:5346)
E/AndroidRuntime( 5200): 	at android.view.ViewRootImpl$AsyncInputStage.apply(ViewRootImpl.java:5535)
E/AndroidRuntime( 5200): 	at android.view.ViewRootImpl$InputStage.deliver(ViewRootImpl.java:5319)
E/AndroidRuntime( 5200): 	at android.view.ViewRootImpl$InputStage.onDeliverToNext(ViewRootImpl.java:5372)
E/AndroidRuntime( 5200): 	at android.view.ViewRootImpl$InputStage.forward(ViewRootImpl.java:5338)
E/AndroidRuntime( 5200): 	at android.view.ViewRootImpl$InputStage.apply(ViewRootImpl.java:5346)
E/AndroidRuntime( 5200): 	at android.view.ViewRootImpl$InputStage.deliver(ViewRootImpl.java:5319)
E/AndroidRuntime( 5200): 	at android.view.ViewRootImpl$InputStage.onDeliverToNext(ViewRootImpl.java:5372)
E/AndroidRuntime( 5200): 	at android.view.ViewRootImpl$InputStage.forward(ViewRootImpl.java:5338)
E/AndroidRuntime( 5200): 	at android.view.ViewRootImpl$AsyncInputStage.forward(ViewRootImpl.java:5511)
E/AndroidRuntime( 5200): 	at android.view.ViewRootImpl$ImeInputStage.onFinishedInputEvent(ViewRootImpl.java:5672)
E/AndroidRuntime( 5200): 	at android.view.inputmethod.InputMethodManager$PendingEvent.run(InputMethodManager.java:3065)
E/AndroidRuntime( 5200): 	at android.view.inputmethod.InputMethodManager.invokeFinishedInputEventCallback(InputMethodManager.java:2608)
E/AndroidRuntime( 5200): 	at android.view.inputmethod.InputMethodManager.finishedInputEvent(InputMethodManager.java:2599)
E/AndroidRuntime( 5200): 	at android.view.inputmethod.InputMethodManager$ImeInputEventSender.onInputEventFinished(InputMethodManager.java:3042)
E/AndroidRuntime( 5200): 	at android.view.InputEventSender.dispatchInputEventFinished(InputEventSender.java:143)
E/AndroidRuntime( 5200): 	at android.os.MessageQueue.nativePollOnce(Native Method)
E/AndroidRuntime( 5200): 	at android.os.MessageQueue.next(MessageQueue.java:336)
E/AndroidRuntime( 5200): 	at android.os.Looper.loop(Looper.java:174)
E/AndroidRuntime( 5200): 	at android.app.ActivityThread.main(ActivityThread.java:7397)
E/AndroidRuntime( 5200): 	at java.lang.reflect.Method.invoke(Native Method)
E/AndroidRuntime( 5200): 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
E/AndroidRuntime( 5200): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:935)
I/Process ( 5200): Sending signal. PID: 5200 SIG: 9
Lost connection to device.

Exited.
```

</details>

--- a/engine/src/flutter/shell/platform/android/io/flutter/embedding/android/KeyEmbedderResponder.java
+++ b/engine/src/flutter/shell/platform/android/io/flutter/embedding/android/KeyEmbedderResponder.java
@@ -183,7 +183,7 @@ public class KeyEmbedderResponder implements KeyboardManager.Responder {
               postSynchronize.add(
                   () ->
                       synthesizeEvent(
-                          false, key.logicalKey, key.physicalKey, event.getEventTime()));
+                          false, key.logicalKey, eventPhysicalKey, event.getEventTime()));
             }
             preEventStates[keyIdx] = nowStates[keyIdx];
             postEventAnyPressed = true;

--- a/engine/src/flutter/shell/platform/android/test/io/flutter/embedding/android/KeyboardManagerTest.java
+++ b/engine/src/flutter/shell/platform/android/test/io/flutter/embedding/android/KeyboardManagerTest.java
@@ -1669,6 +1669,68 @@ public class KeyboardManagerTest {
     calls.clear();
   }
 
+  // Regression test for https://github.com/flutter/flutter/issues/164626.
+  @Test
+  public void synchronizeModifiersForZeroedScanCodeOnRepeatEvent() {
+    // Test if ShiftLeft can be correctly synchronized during down events of
+    // ShiftLeft that have 0 for their metaState and 0 for their scanCode.
+    final KeyboardTester tester = new KeyboardTester();
+    final ArrayList<CallRecord> calls = new ArrayList<>();
+
+    tester.recordEmbedderCallsTo(calls);
+    tester.respondToTextInputWith(true); // Suppress redispatching
+
+    // Test: repeat event when the meta state is 0 and scanCode is 0.
+    final KeyEvent shiftLeftKeyEvent =
+        new FakeKeyEvent(ACTION_DOWN, 0, KEYCODE_SHIFT_LEFT, 1, '\0', 0);
+    // Compute physicalKey in the same way as KeyboardManager.getPhysicalKey.
+    final Long shiftLeftPhysicalKey = KEYCODE_SHIFT_LEFT | KeyboardMap.kAndroidPlane;
+
+    assertEquals(tester.keyboardManager.handleEvent(shiftLeftKeyEvent), true);
+    assertEquals(calls.size(), 2);
+    assertEmbedderEventEquals(
+        calls.get(0).keyData,
+        Type.kDown,
+        shiftLeftPhysicalKey,
+        LOGICAL_SHIFT_LEFT,
+        null,
+        false,
+        DeviceType.kKeyboard);
+    assertEmbedderEventEquals(
+        calls.get(1).keyData,
+        Type.kUp,
+        shiftLeftPhysicalKey,
+        LOGICAL_SHIFT_LEFT,
+        null,
+        true,
+        DeviceType.kKeyboard);
+    calls.clear();
+
+    // Similar check for AltLeft.
+    final KeyEvent altLeftKeyEvent = new FakeKeyEvent(ACTION_DOWN, 0, KEYCODE_ALT_LEFT, 1, '\0', 0);
+    final Long altLeftPhysicalKey = KEYCODE_ALT_LEFT | KeyboardMap.kAndroidPlane;
+
+    assertEquals(tester.keyboardManager.handleEvent(altLeftKeyEvent), true);
+    assertEquals(calls.size(), 2);
+    assertEmbedderEventEquals(
+        calls.get(0).keyData,
+        Type.kDown,
+        altLeftPhysicalKey,
+        LOGICAL_ALT_LEFT,
+        null,
+        false,
+        DeviceType.kKeyboard);
+    assertEmbedderEventEquals(
+        calls.get(1).keyData,
+        Type.kUp,
+        altLeftPhysicalKey,
+        LOGICAL_ALT_LEFT,
+        null,
+        true,
+        DeviceType.kKeyboard);
+    calls.clear();
+  }
+
   @Test
   public void normalCapsLockEvents() {
     final KeyboardTester tester = new KeyboardTester();

